legend-sdlc-server/src/main/java/org/finos/legend/sdlc/server/gitlab/api/GitlabWorkflowJobApi.java

```diff
@@ -0,0 +1,326 @@
+        LegendSDLCServerException.validateNonNull(versionId, "versionId may not be null");
+        return new GitLabWorkflowJobAccessContext(projectId)
+        {
+            @Override
+            protected String getRef()
+            {
+                return buildVersionTagName(versionId);
+            }
+
+            @Override
+            protected String getRefInfoForException()
+            {
+                return "version " + versionId.toVersionIdString() + " of project " + projectId;
+            }
+        };
+    }
+
+    private abstract class GitLabWorkflowJobAccessContext implements WorkflowJobAccessContext
+    {
+        private final GitLabProjectId projectId;
+
+        private GitLabWorkflowJobAccessContext(GitLabProjectId projectId)
+        {
+            this.projectId = projectId;
+        }
+
+        private GitLabWorkflowJobAccessContext(String projectId)
+        {
+            this(parseProjectId(projectId));
+        }
+
+        @Override
+        public WorkflowJob getWorkflowJob(String workflowId, String workflowJobId)
+        {
+            int jobId = parseIntegerIdIfNotNull(workflowJobId);
+            Job job;
+            try
+            {
+                job = withRetries(() -> getGitLabApi(this.projectId.getGitLabMode()).getJobApi().getJob(this.projectId.getGitLabId(), jobId));
+            }
+            catch (Exception e)
+            {
+                throw buildException(e,
+                        () -> "User " + getCurrentUser() + " is not allowed to access workflow job " + workflowJobId + " in " + getRefInfoForException(),
+                        () -> "Unknown workflow job in " + getRefInfoForException() + ": " + workflowJobId,
+                        () -> "Error getting workflow job " + workflowJobId + " in " + getRefInfoForException());
+            }
+
+            if (!getRef().equals(job.getRef()))
+            {
+                throw new LegendSDLCServerException("Unknown workflow job in " + getRefInfoForException() + ": " + workflowJobId, Response.Status.NOT_FOUND);
+            }
+
+            if (!workflowId.equals(toStringIfNotNull(job.getPipeline().getId())))
+            {
+                throw new LegendSDLCServerException("Unknown workflow job in " + getRefInfoForException() + ", workflow " + workflowId + ": " + workflowJobId, Response.Status.NOT_FOUND);
+            }
+
+            return fromGitLabJob(this.projectId.toString(), job);
+        }
+
+        @Override
+        public List<WorkflowJob> getWorkflowJobs(String workflowId, Iterable<WorkflowJobStatus> statuses)
+        {
+            try
+            {
+                int pipelineId = parseIntegerIdIfNotNull(workflowId);
+                JobApi jobApi = getGitLabApi(this.projectId.getGitLabMode()).getJobApi();
+                List<Job> jobs = withRetries(() -> jobApi.getJobsForPipeline(this.projectId.getGitLabId(), pipelineId));
```

Comment:
This looks like it's a limitation of the gitlab4j API. We should be getting a Pager here, not a List. That's going to cause problems for pipelines with a large number of jobs (where the number of jobs exceeds the maximum number of items per page). We might want to contribute an improvement to gitlab4j to fix this.